src/app/util/types_stub.h

```diff
@@ -538,16 +502,13 @@ typedef struct
     chip::ClusterId clusterId;
     /** The endpoint on the remote node (specified by \c identifier). */
     chip::EndpointId remote;
-    /** A 64-bit destination identifier.  This is either:
-     * - The destination chip::NodeId, for unicasts.
-     * - A multicast ChipGroupId, for multicasts.
-     * Which one is being used depends on the type of this binding.
-     */
-    union
-    {
-        chip::NodeId nodeId;
-        chip::GroupId groupId;
-    };
+
+    /** The exchange associated with the binding. Messages are send or received
```

Comment:
No.  No.  Bindings have node or group ids.  The CANNOT have a transaction or exchange associated with them.  Bindings are long-lived data structures that survive reboot and the like.  Transactions/exchanges are short-lived.  Furthermore, there might be multiple transactions at the same time that are all based on the same binding!

Reply:
In this case we need to translate these persistent bindings into active binding, where `ExchangeContext` are stored. We can't send message only given node ids.

Reply:
OK, so a few things.  Maybe I completely misunderstand the scope of an `ExchangeContext`?  My understanding was that there was one `ExchangeContext` per exchange, with it being created when the exchange starts and destroyed when the exchange ends.  Is that correct?

If that is correct, then the right flow would seem to be that when the data model wants to start a new transaction (which corresponds to an exchange) it needs to _at that point_ create an `ExchangeContext` and send via that context.

Is my understanding of the interaction between exchanges and `ExchangeContext` incorrect?  If not, is my understanding that a new exchange is cheap and quick to set up (which a bunch of the existing specification text is based on) incorrect?  If we have to do mDNS discovery for every single exchange setup, that doesn't seem obviously acceptable for the way the current interaction model spec proposes to use exchanges.

My understanding from the last time I spoke to folks about the APIs data model should deal with in terms of interacting with lower layers was that it would go about as follows:

1. For incoming messages and their replies, there would be a representation of the exchange that came with the message that could be used to send a reply.  Pretty straightforward.
2. For message origination, the data model would hand a node id or group id to the lower layers and get back a representation of a newly created exchange.  It would then be responsible for managing the lifetime of that exchange and send messages pertaining to one data model transaction through it.  When the transaction is done, the exchange would be released.  Any details about secure channel establishment, determining and caching mappings from node ids to ips, etc, would be an implementation detail of the message layer.

Has the API contract changed since then?  If so, can you please describe the new API contract?

> We can't send message only given node ids.

For an outgoing message, given the current spec setup, we _must_ on the data model level, be able to do precisely that, by creating a new exchange from that node id (which might require establishment of a secure channel if one is not already established, of course, but that's a message layer implementation detail) and then sending messages via the exchange.

Reply:
Yes, you are absolutely right. exchanges are short-live and cheap. But lots of prerequisite is needed. If it doesn't require an exchange, then it still need a structure like `Cached<PeerAddress, KeyId, SecureSessionState>` to send messages.

If we leave the caching mechanism to underlayer implementations totally, it will probably fail to send several reports when something happens, like device reboots, peer IP address changes, session key rotates. With the `ExchangeContext`, the data model is able to acknowledge `OnConnectionClosed` event, and re-establish the channel as soon as possible.

Reply:
Maybe that's the heart of the disagreement about how this should work?  My understanding was that channels are expensive to track and a node can potentially need way more channels than it can support in-flight at once, so channels are set up on an as-needed basis, not constantly maintained.  We definitely need to sort this out, but it seems to be out of the scope of this PR as well...  

In any case, it seems to me that what ends up happening when we want to do a new exchange is that we should be able to ask the underlying layer for a new exchange.  Sometimes that will require channel establishment, because we can't keep all channels open all the time, in my understanding, so whatever APIs and data flows we design around this need to be able to handle that.  If there is channel caching and keeping alive going on, it should be separate from the basic mechanism of "ok, we want to send to this node id; look up the channel for it or create one if there isn't one yet"....

If I am totally wrong and the intent _is_ in fact to keep all possible outgoing channels active at all times, then obviously that changes our options for the design of the sending APIs in data model.